src/resolve/resolved-dns-dnssec.c

```diff
@@ -613,6 +711,9 @@ int dnssec_verify_rrset(
         gcry_md_hd_t md = NULL;
         int r, md_algorithm;
         size_t k, n = 0;
+        size_t sig_size = 0;
+        _cleanup_free_ char *sig_data = NULL;
```

Comment:
`free` has to be called after `fclose`. The GCC documentation for the `cleanup` attribute does not state in which order the functions are called. I only found a Google+ post by Lennart Poettering that claims that they are called “in order”, presumably in the reverse order of declaration. I would like to avoid that we depend on undefined behaviour as the wrong order caused a memory corruption when I tested it.

Reply:
well, we do that all over the place already by ordering the declarations properly, so by doing this you are not going to break anything new...

Moreover the gcc destructor stuff is pretty much identical to the C++ destructor logic, and exposes the same semantics. And for C++ the standard is very explicit about the destruction order, see https://stackoverflow.com/a/2254306